### PR TITLE
Add protocol update action

### PR DESCRIPTION
**Pull request by diehuxx**
The `update` action allows an actor to update an existing record. This contrasts with the `write` action which allows only the creator of a record to update it. In other words:
1. `write` allows a DID to create a record and update the record they have created
2. `update` allows a DID to update a record, regardless of initial author

### CODE REVIEW

**Comment by alanhkarp** [comment]
Is this distinction necessary?  Can't you just use capabilities for this?

Say that Alice creates some record.  As I understand it, the DWN node can't create a capability to return to her, but it does know she is the owner.  That means Alice can create a "write" capability that she can use.  She can also delegate that capability to someone else.

The upshot is that you only need one "write" method at the expense of the owner having to create a root capability.  An added advantage is that Alice won't be vulnerable to a confused deputy attack as she would be if she used her ambient authority.

File: tests/vectors/protocol-definitions/author-update.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "protocol": "http://recipient-update-protocol.xyz",
```
Comment (thehenrytsai) [inline tests/vectors/protocol-definitions/author-update.json:2]: Super nit:
```suggestion
  "protocol": "http://author-update-protocol.xyz",
```
Also the protocol is a bit weird in that it only allows the owner to comment, but fair enough given this is all superficial anyways.

File: src/core/protocol-authorization.ts [hunk 2]
```diff
@@ -333,17 +328,42 @@ export class ProtocolAuthorization {
...
   private static async verifyAllowedActions(
+    tenant: string,
     incomingMessage: RecordsRead | RecordsWrite,
     inboundMessageRuleSet: ProtocolRuleSet,
     ancestorMessageChain: RecordsWriteMessage[],
+    messageStore: MessageStore,
   ): Promise<void> {
     const incomingMessageMethod = incomingMessage.message.descriptor.method;
-    const inboundMessageAction = methodToAllowedActionMap[incomingMessageMethod];
+    const inboundMessageActions = await ProtocolAuthorization.getInboundMessageActions(tenant, incomingMessage, messageStore);
```
Comment (thehenrytsai) [inline src/core/protocol-authorization.ts:366]: Naming is a bit vague to me, consider renaming to perhaps `getActionsSeekingARuleMatch()` or `getPotentiallyAllowedActions()` or `getActionCandidates()`

```suggestion
    const actionsSeekingARuleMatch = await ProtocolAuthorization.getActionsSeekingARuleMatch(tenant, incomingMessage, messageStore);
```

**Review by thehenrytsai — DISMISSED** [review]
LGTM and test coverage is super thorough!! Just a few superficial comment.

**Comment by diehuxx** [comment]
> Is this distinction necessary? Can't you just use capabilities for this?
> 
> Say that Alice creates some record. As I understand it, the DWN node can't create a capability to return to her, but it does know she is the owner. That means Alice can create a "write" capability that she can use. She can also delegate that capability to someone else.
> 
> The upshot is that you only need one "write" method at the expense of the owner having to create a root capability. An added advantage is that Alice won't be vulnerable to a confused deputy attack as she would be if she used her ambient authority.

@csuwildcat Any thoughts?

File: src/interfaces/records-write.ts [hunk 3]
```diff
@@ -602,6 +604,17 @@ export class RecordsWrite {
     return (entryId === this.message.recordId);
   }
 
+  public async authoredByInitialRecordAuthor(tenant: string, messageStore: MessageStore): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply src/interfaces/records-write.ts]: Added summary doc and updated method name

File: tests/vectors/protocol-definitions/author-update.json (hunk 1 above)
Comment (diehuxx) [reply tests/vectors/protocol-definitions/author-update.json]: Updated string. You're right, it's not the most realistic example. This is mostly a test fixture than an insightful example to read.

File: src/core/protocol-authorization.ts (hunk 2 above)
Comment (diehuxx) [reply src/core/protocol-authorization.ts]: Updated

**Comment by LiranCohen** [comment]
> That means Alice can create a "write" capability that she can use. She can also delegate that capability to someone else.

Yes, you could technically build the same type of behavior with delegation for other aspects of the protocol rules, but it would be difficult to enforce these as a 'protocol' with other implementations.

Say you have a protocol with the following rules:
1) anyone can create a `post`.
2) anyone can add a `post/tag` to existing posts.
3) the `did` who created the `post` can edit any `post/tag` relating to the `post` they created.

Instead of each implementation of this social protocol front-end app explicitly delegating the write capability of the `post/tag` they created to the author of the `post`, the protocol rules handle this.

**Review by LiranCohen — APPROVED** [review]
LGTM!

**Comment by alanhkarp** [comment]
@LiranCohen 

> @alanhkarp
> 
> > That means Alice can create a "write" capability that she can use. She can also delegate that capability to someone else.
> 
> Yes, you could technically build the same type of behavior with delegation for other aspects of the protocol rules, but it would be difficult to enforce these as a 'protocol' with other implementations.
> 
> Say you have a protocol with the following rules:
> 
> 1. anyone can create a `post`.
> 2. anyone can add a `post/tag` to existing posts.

In that case, allow the creator of the `post` to specify that no capability is needed to add a `post/tag` to the post. 

> 3. the `did` who created the `post` can edit any `post/tag` relating to the `post` they created.
> 

Using the `did` of the creator of the `post` is effectively using an access control list.  (The same goes for using Roles as I pointed out in another issue.)  It's not as bad a a pure ACL system, since Alice can create a capability when she wants to delegate some of her rights.  The risk, though, is a confused deputy vulnerability.

**Comment by csuwildcat** [comment]
@nidhinahar looks good to me

**Review by thehenrytsai — APPROVED** [review]

**Pull request merged**